lib/mayaUsd/fileio/shading/shadingModeExporterContext.cpp

```diff
@@ -374,21 +373,25 @@ static UsdPrim _GetMaterialParent(
             }
         }
     }
+    return commonAncestor;
+}
 
-    if (commonAncestor.IsEmpty()) {
-        return UsdPrim();
-    }
+static UsdPrim _GetMaterialParent(
+    const UsdStageRefPtr&                                    stage,
+    const std::string&                                       defaultPrim,
+    const TfToken&                                           materialsScopeName,
+    const UsdMayaShadingModeExportContext::AssignmentVector& assignments)
+{
+    SdfPath shaderExportLocation = _GetCommonAncestor(stage, assignments);
 
-    if (commonAncestor == SdfPath::AbsoluteRootPath()) {
-        return stage->GetPseudoRoot();
-    }
+    if (shaderExportLocation.IsEmpty())
+        shaderExportLocation = SdfPath::AbsoluteRootPath();
 
-    SdfPath shaderExportLocation = commonAncestor;
-    while (!shaderExportLocation.IsRootPrimPath()) {
-        shaderExportLocation = shaderExportLocation.GetParentPath();
-    }
```

Comment:
Hi @pierrebai-adsk - this change broke a few of our export tests since the logic to find the MaterialParent path is significantly different. Prior to this change, all materials used under prim `/Foo` would be found in `/Foo/Looks`, but now they will now either be included under the root prim ('/Looks', as a sibling to '/Foo') or will be located under the specific child prim in which the material binding is found ('/Foo/Geom/Child/Looks'), depending on whether or not we set the new export option.

This violates a few internal assumptions we have in our pipeline, and I'm wondering if it was an intentional decision to change existing and established export behavior. If so we may need to either patch this change out of our build, or introduce another export option that would preserve the behavior our pipeline expects.

thanks

Reply:
Also, just so I'm sure I understand what the intended behavior is:

- If I specify `exportMaterialUnderPrim = true` then the materialParent should be `/Path/to/materialbinding`
- If I specify only a `defaultPrim` in my export args, then the parent should be the default prim path
- If I specify both, should the parent be `/Path/to/materialbinding + /defaultPrimPath`?

Reply:
It seems like I might be able to recover the established behavior if I leave `exportMaterialUnderPrim` and I specify `defaultPrim` in my export args, but we don't typically specify defaultPrim in our export args - we allow the export process to determine it for us here (I think):
http://grok/xref/mainline/ext/mayapkg/maya_open_source/maya-usd/lib/mayaUsd/fileio/jobs/writeJob.cpp?r=2328671#565

Unfortunately we don't have access to that defaultPrim data at the time of this call

Reply:
Hi @dj-mcg 

Yes, the changes were intentional, the designer wanted to make the location of materials more predictable. The materials now are generally under the default prim if there is one, under the root if there is none.

The exportMaterialUnderPrim was added so that our edit-as-Maya / merge-to-USD workflow work, since that assumed everything was under the edited prim.

As you can see in this PR, we also had to adjust multiple tests to take into account the new location. For your case, the export-under-prim flag does not do the job of restoring the behavior that you expect?

Reply:
Hey @dj-mcg, in the meantime while we look at this issue on our end, I'll invite you to take a look at the first part of github #3674 where I describe the existing behavior versus the new behavior in terms of where materials end up.   Lemme know if you have any thoughts/concerns, thanks!

Reply:
Hi @scottrenaud - thanks for the link. The proposal made complete sense to me, but my only concern with the implementation is that the Default Prim needs to be explicitly specified in exportArgs, and it does not use the same logic of falling back to the "UsdRootPrim" if this export arg is empty as the code in `UsdMaya_WriteJob::_FinishWriting`. Perhaps this was an intentional decision and I made an incorrect assumption when originally reading the proposal - If so I wonder if we can add this behavior via another exportArg?

Reply:
Hey @dj-mcg, just to follow up here, as mentioned in https://github.com/Autodesk/maya-usd/pull/3810#discussion_r1631464040 you're shining light on the fact that this is a gap, we'll create a ticket for this internally and discuss it with the team, thanks!